src/Compilers/VisualBasic/Portable/Symbols/UsedAssemblies.vb

```diff
@@ -0,0 +1,211 @@
+
+                If Not seenErrors Then
+                    diagnostics.DiagnosticBag.Clear()
+                    ' PROTOTYPE(UsedAssemblyReferences): Might want to suppress nullable analysis for this call.
+                    ' PROTOTYPE(UsedAssemblyReferences): Ensure we don't trigger any analyzers during the GetDiagnosticsForAllMethodBodies call.
+                    GetDiagnosticsForAllMethodBodies(hasDeclarationErrors:=False, diagnostics, doLowering:=True, cancellationToken)
+                    seenErrors = diagnostics.HasAnyErrors()
+
+                    If Not seenErrors Then
+                        AddUsedAssemblies(diagnostics.DependenciesBag)
+                    End If
+                End If
+
+                CompleteTheSetOfUsedAssemblies(seenErrors, cancellationToken)
+
+                diagnostics.DiagnosticBag.Free()
+            End If
+
+            Return _lazyUsedAssemblyReferences
+        End Function
+
+
+        Private Sub AddUsedAssembly(dependency As AssemblySymbol, stack As ArrayBuilder(Of AssemblySymbol))
+            If AddUsedAssembly(dependency) Then
+                stack.Push(dependency)
+            End If
+        End Sub
+
+        Private Sub AddReferencedAssemblies(assembly As AssemblySymbol, includeMainModule As Boolean, stack As ArrayBuilder(Of AssemblySymbol))
+            For i As Integer = If(includeMainModule, 0, 1) To assembly.Modules.Length - 1
+                For Each dependency In assembly.Modules(i).ReferencedAssemblySymbols
+                    AddUsedAssembly(dependency, stack)
+                Next
+            Next
+        End Sub
+
+        Private Sub CompleteTheSetOfUsedAssemblies(seenErrors As Boolean, cancellationToken As CancellationToken)
+
+            If _usedAssemblyReferencesFrozen OrElse Volatile.Read(_usedAssemblyReferencesFrozen) Then
+                Return
+            End If
+
+            If seenErrors Then
+                ' Add all referenced assemblies
+                For Each assembly As AssemblySymbol In SourceModule.ReferencedAssemblySymbols
+                    AddUsedAssembly(assembly)
+                Next
+            Else
+                ' Assume that all assemblies used by the added modules are also used
+                For i As Integer = 1 To SourceAssembly.Modules.Length - 1
+                    For Each dependency In SourceAssembly.Modules(i).ReferencedAssemblySymbols
+                        AddUsedAssembly(dependency)
+                    Next
+                Next
+
+                ' Assume that all assemblies used by the used assemblies are also used
+                ' This, for example, takes care of including facade assemblies that forward types around.
+                If _lazyUsedAssemblyReferences IsNot Nothing Then
+                    SyncLock _lazyUsedAssemblyReferences
+                        If _usedAssemblyReferencesFrozen OrElse Volatile.Read(_usedAssemblyReferencesFrozen) Then
```

Comment:
>_usedAssemblyReferencesFrozen [](start = 27, length = 29)

Can we do a read of this outside the lock?

Reply:
> Can we do a read of this outside the lock?

We do this at the beginning of the function and I think it make sense to check after the lock is acquired too in order to avoid doing the work if it was done while we were waiting for the lock.

---
In reply to: [369262446](https://github.com/dotnet/roslyn/pull/40987#discussion_r369262446) [](ancestors = 369262446)